scipy/interpolate/src/_fitpackmodule.c

```diff
@@ -191,14 +191,14 @@ fitpack_bispev(PyObject *dummy, PyObject *args)
     ny = PyArray_DIMS(ap_ty)[0];
     mx = PyArray_DIMS(ap_x)[0];
     my = PyArray_DIMS(ap_y)[0];
-    mxy = mx*my;
-    if (my != 0 && mxy/my != mx) {
+    if (my != 0 && INTPTR_MAX/my < mx) {
```

Comment:
I see that this works, but I would love any explanation as to why (by my algebra) it is always true (in infinite precision) that `INTPTR_MAX - my * mx >= INTPTR_MAX % mx` for suitable `mx`, `my`.  I see why that would be false where `INTPTR_MAX < my * mx`, but not why it will always be true when `INTPTR_MAX >= my * mx`.

Reply:
I have to agree to this comment completely...

Reply:
Maybe we need this: https://github.com/numpy/numpy/blob/a87ee593c3f5da5dcb42a8033ecbab0f0c9b7903/numpy/core/src/common/npy_extint128.h#L40-L56 (mentioned in the original issue)

Reply:
The intuitive way to see this is that `INTPTR_MAX - INTPTR_MAX % mx` is the largest multiple of `mx`  that is less than or equal to `INTPTR_MAX`. So your condition is equivalent to 
```
my * mx <= largest valid multiple of mx
```

A more formal way to see this: 

Let `INTPTR_MAX / mx = a` and `INTPTR_MAX % mx = b`, so that `INTPTR_MAX = a * mx + b`. Since all initial quantities are dimension length and so non-negative, we know that `a` and `b` are `>= 0` as well.

Now rewriting your expression we get
```
a*mx + b - my*mx >= b
a*mx >= my*mx
a >= my                     (since mx > 0)
my <= a
my <= INTPTR_MAX/mx
```
Which is the negative of the test above with `mx` and `my` swapped, but multiplication is symmetric so that doesn't matter. 

Now, what happens if `my <= a`? 
```
my <= a
mx*my <= a*mx               (since mx > 0)
mx*my <= a*mx + b           (since b >= 0)
mx*my <= INTPTR_MAX
```
So `my <= a` is sufficient to show that `mx*my <= INTPTR_MAX`.

Now what happens if `my > a`?
```
my > a
my >= a + 1                 (since my and a are integral)
mx*my >= (a+1)*mx
mx*my >= a*mx + mx
mx*my > a*mx + b            (since b < mx)
mx*my > INTPTR_MAX
```
Thus `my > a` means `mx*my > INTPTR_MAX`, and so `my <= a` is necessary for `mx*my` to be valid. 

Together these show `my <= INTPTR_MAX / mx` is necessary and sufficient to show that `mx*my <= INTPTR_MAX`.

Reply:
@peterbell,  thanks, great.

Reply:
another question: macOS test (meson) was failing - because of these changes? Otherwise it would be good to accept this PR.

Reply:
It seems to me, that `INT_MAX` (defined by `pyport.h`) and `INTPTR_MAX` (`stdint.h`) is the same?

Reply:
Python 32-bit: `include/pyport.h`:`#define INT_MAX 2147483647`
mingw-w64 `stdint.h`: `INTPTR_MAX == INT32_MAX == 2147483647`

In the 64-bit case: `INTPTR_MAX == INT64_MAX` is much larger than `INT_MAX`. However, the fitpack issues doesn't shows up with scipy-meson windows 64-bit.

Reply:
And what is `F_INT` on your windows 64 bit build?

Reply:
I'm not aware of `F_INT`, but `__SIZEOF_INT__` it is defined like that: `#define __SIZEOF_INT__ 4` for both 32 and 64 bit

Reply:
`F_INT` is what matters here though, because it determines when there is a risk of fortran overflowing in its calculation.

Reply:
https://github.com/scipy/scipy/blob/main/scipy/interpolate/src/_fitpackmodule.c#L35

Reply:
I **didn't** use `-fdefault-integer-8`, so the integer size is 4.